File: divideAndConquer/searchMatrix.ts
// binary search basically eliminates 1/4 of the matrix at each iteration
function searchMatrixBinary(matrix: number[][], target: number): boolean {
  let ROWS = matrix.length;
  let COLS = matrix[0].length;

  function search(
    startRow: number,
    endRow: number,
    startCol: number,
    endCol: number
  ): boolean {
    if (startRow > endRow || startCol > endCol) {
      return false;
    }

    // 'divide'
    let midRow = Math.floor((startRow + endRow) / 2);
    let midCol = Math.floor((startCol + endCol) / 2);
    let midNum = matrix[midRow][midCol];

    if (midNum === target) {
      return true;
    }

    if (midNum > target) {
      // target is smaller than our mid partition
      return (
        // 'conquer + combine'
        search(startRow, midRow - 1, startCol, endCol) || // top half
        search(midRow, endRow, startCol, midCol - 1) // left half
      );
    } else {
      // target is greater than our mid partition
      return (
        // 'conquer + combine'
        search(startRow, midRow, midCol + 1, endCol) || // right half
        search(midRow + 1, endRow, startCol, endCol) // bottom half
      );
    }
  }

  return search(0, ROWS - 1, 0, COLS - 1);
}

// this traverses in a diagonal fashion across the whole matrix, less efficient than binSearch method
function searchMatrix(matrix: number[][], target: number): boolean {
  let ROWS = matrix.length;
  let COLS = matrix[0].length;
  let exists = false;

  search(0, COLS - 1);

  function search(currentRow: number, currentCol: number) {
    let currentNum = matrix[currentRow][currentCol];

    if (currentNum === target) {
      exists = true;
      return;
    }

    if (target < currentNum && currentCol > 0) {
      search(currentRow, currentCol - 1);
    } else if (currentRow < ROWS - 1) {
      search(currentRow + 1, currentCol);
    }
  }

  return exists;
}

/*

Write an efficient algorithm that searches for a value target in an m x n integer matrix matrix. This matrix has the following properties:

Integers in each row are sorted in ascending from left to right.
Integers in each column are sorted in ascending from top to bottom.

-

Input: matrix = [[1,4,7,11,15],[2,5,8,12,19],[3,6,9,16,22],[10,13,14,17,24],[18,21,23,26,30]], target = 5
Output: true

-

start at [0, 0]

5 != 1 and 5 > 4, so go right

5 != 4 but 5 < 7, go down

5 === 5, return true

-

Input: matrix = [[1,4,7,11,15],[2,5,8,12,19],[3,6,9,16,22],[10,13,14,17,24],[18,21,23,26,30]], target = 20

20 != 1 and 20 > 4, right
20 != 1 and 20 > 7, right
20 != 1 and 20 > 11, right
20 != 1 and 20 > 15, but no more right elements - so go down

-

generalize more...

if target === current num, return true

we can search right or down at each instance:

  if target > num and right cell exists, search right

  if target > num and bottom cell exists, search bottom

*/

File: binSearch/searchMatrix.py
class Solution:
  def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
      rows = len(matrix)
      cols = len(matrix[0]) if rows > 0 else 0
      s = 0
      e = rows * cols - 1

      while s <= e:
          m = (s + e) // 2
          mid_val = matrix[m // cols][m % cols]

          if mid_val == target:
              return True
          elif mid_val < target:
              s = m + 1
          else:
              e = m - 1

      return False


class Solution1:
  def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
    s = 0
    e = len(matrix) - 1
    target_row_idx = None

    # find row where target would be
    if len(matrix) == 1:
      target_row_idx = 0
    else:
      while s <= e:
        m = (s + e) // 2

        if target >= matrix[m][0] and target <= matrix[m][-1]:
          target_row_idx = m
          break
        elif target > matrix[m][-1]:
          s = m + 1
        elif target < matrix[m][0]:
          e = m - 1

    # check for target in the target row
    if target_row_idx is not None:
      s = 0
      e = len(matrix[target_row_idx]) - 1

      while s <= e:
        m = (s + e) // 2

        if target == matrix[target_row_idx][m]:
          return True
        elif target < matrix[target_row_idx][m]:
          e = m - 1
        elif target > matrix[target_row_idx][m]:
          s = m + 1

    return False

'''

[[1,3,5,7],[10,11,16,20],[23,30,34,60]]
target = 3
output = true

-

- sorted and O(log(m*n)) suggests bin search

-

start by partitioning rows until we have one row, then partition by that row's elements

-

if target >= midrow first el and target <= midrow last el:

  search this row

  how to go from rows to elements of rows?

else if target > midrow last el

  search right

else if target < midrow first el

  search left

'''
        